### PR TITLE
GenericCopyUtil refactoring for improving I/O performance

### PR DESCRIPTION
**Pull request by TranceLove**
Following discussion at #1572 I made an attempt to improve GenericCopyUtil on I/O operations.

* Unified all file copy operations to use NIO (under the hood, using delegation)
* Consequently, removed `CustomReadableByteChannel` and increment `ServiceWatcherUtil.position` directly
* Increased stream buffer size of SFTP and SMB to 64KB. This seems to be the better buffer size that increased network I/O copy performance without risk crashing the app.

Espresso and Robolectric tests for the refactoring are available. Sorry for the `DummyFileGenerator` in `src/main`, but it's the only way to make the class available to both Espresso and Robolectric tests - it had been guarded against main app code with `@VisibleForTesting` annotation.

Tested using Oneplus 2 running Slim7 (7.1.2) and Sony Xperia Z1 Compact running CM13 (6.0.1).

Known issue: `CopyService` was seen not able to stop after copying multiple files, i.e. the notification pull down did not go away after finished copying, on Sony Xperia Z1 Compact running CM13 (6.0.1).

### CODE REVIEW

File: app/src/main/java/com/amaze/filemanager/utils/files/GenericCopyUtil.java [hunk 1]
```diff
@@ -325,129 +334,94 @@ public void copy(HybridFileParcelable sourceFile, HybridFile targetFile) throws
...
-                ServiceWatcherUtil.position += bytes;
-                return bytes;
+        buffer.flip();
+        while(buffer.hasRemaining())
+            to.write(buffer);
 
-            } else
-                return 0;
-        }
+        from.close();
```
Comment (VishalNehra) [inline app/src/main/java/com/amaze/filemanager/utils/files/GenericCopyUtil.java]: No need to close channels as they are being done in finally block of ````startCopy````

File: app/src/main/java/com/amaze/filemanager/utils/files/GenericCopyUtil.java [hunk 2]
```diff
@@ -325,129 +334,94 @@ public void copy(HybridFileParcelable sourceFile, HybridFile targetFile) throws
...
+        to.close();
+    }
 
-        @Override
-        public boolean isOpen() {
-            return byteChannel.isOpen();
+    @VisibleForTesting
+    void doCopy(@NonNull ReadableByteChannel from, @NonNull FileChannel to) throws IOException{
```
Comment (VishalNehra) [inline app/src/main/java/com/amaze/filemanager/utils/files/GenericCopyUtil.java]: Do we really need this method? AFAIK, ````FileChannel```` implements both ````WritableByteChannel```` and ````ReadableByteChannel````.

Comment (TranceLove) [reply app/src/main/java/com/amaze/filemanager/utils/files/GenericCopyUtil.java]: `transferFrom` is implemented in FileChannel.

https://developer.android.com/reference/java/nio/channels/FileChannel#transferFrom(java.nio.channels.ReadableByteChannel,%20long,%20long)

File: app/src/main/java/com/amaze/filemanager/utils/files/GenericCopyUtil.java (hunk 1 above)
Comment (TranceLove) [reply app/src/main/java/com/amaze/filemanager/utils/files/GenericCopyUtil.java]: Unfortunately, without closing the channels `GenericCopyUtilTest.testCopyFile4()` will fail.

Comment (VishalNehra) [reply app/src/main/java/com/amaze/filemanager/utils/files/GenericCopyUtil.java]: I see.

File: app/src/main/java/com/amaze/filemanager/utils/files/GenericCopyUtil.java (hunk 2 above)
Comment (EmmanuelMess) [reply app/src/main/java/com/amaze/filemanager/utils/files/GenericCopyUtil.java]: IMO it is better to leave as much implementation as possible to Android, especially when dealing with the file system.

**Comment by VishalNehra** [comment]
Approved this, but please modfiy the changes I asked for later when you get time @TranceLove

**Comment by TranceLove** [comment]
@VishalNehra you mean dropping `FileChannel`-specific changes in favour of simply using `ByteBuffer`?

**Comment by VishalNehra** [comment]
@TranceLove yes

**Comment by TranceLove** [comment]
Made changes per @VishalNehra requested.

For the record, I refactored `GenericCopyUtilEspressoTest` so it tests doCopy() methods separately instead of grouping them together. Ran the test twice on Oneplus 2 running Slim7 (7.1.2), the was indeed a case where `FileChannel.transferFrom()` ran faster than while loop (as used by `doCopy(ReadableByteChannel from, WritableByteChannel)`, but the other test ran showed the opposite result.

Maybe more sampling can prove @EmmanuelMess 's suggestion is correct, but anyway.

**Comment by TranceLove** [comment]
Two ticks, merging.

**Pull request merged**